Microsoft/react-native-windows
# Fix oacr warnings

Fix all the OACR errors!

ChakraExector.cpp:
* IP ports have been ushorts since forever.

ChakraHelpers.cpp:
* Letting normal type-inference happen in the call to make_unique causes warning 4267 to be raised inside of the implementation of make_unique, where the compiler doesn't seen that the constant value of the `sizeof` expression is much smaller that `UINT_MAX`.

ChakraInstanceArgs.h:
* TCP ports have been ushorts since forever.

ChakraNativeModules.h:
* Need to explicitly cast from size_t to double because that is potentially lossy on x64 platforms (in theory, if result->index were great than 2^53).

ChakraRuntime.cpp:
* IP ports have been ushorts since forever.
* for-loop index variables had the wrong type.
* Chakra uses `int` for array indices, but JSI uses `size_t`. So we need to explicitly cast between them.
* Chakra uses 'unsigned short' for argument counts, but JSI uses `size_t`. So we explicitly cast between them (should probably have error handling here?).
* `ChakraRuntime::JSStringToSTLWString` was implemented in a goofy way with a TODO in there. Yuck. Change it to use `JsGetStringLength` (which returns a count of UTF16 code units) to figured out the right buffer size to use before calling `JsCopyStringUtf16`.

ChakraUtils.cpp:
* Add asserts & explicit casts from `size_t` to `unsigned int`.

ChakraValue.cpp:
* Explicit casts from `size_t` to `int` (in practice an `initiailizer_list` can't be larger than 2^31, but we have asserts anyways.
* `for` loop in `ChakraObject::callAsFunction` had an `unsigned int` index variable, should just be an `int`.
* `count` variables in `ChakraObject::getPropertyNames` and `ChakraObject::toJSONMap` were `size_t`, which is the wrong type. Just use `auto` and `decltype` to let the compiler pick valid types.

Redirects.cpp/h:
* `JsRun`/`JsRunScript`/`JsRunSerialized` had an `unsigned __int64` parameter which ought to be a `uintptr_t` - fixed accordingly.

SandboxBridge.cpp:
* `SandboxJsToNativeBridge::callNativeModules()` needs to explicitly cast from an unsigned `size_t` to a signed `int64_t`. Asserts verify that we don't accidentally make a negative number by having a `size_t` >2^63.

WebSocketIntegrationTest.cpp
* `for` loop index variable should be `size_t`

WebSocketResourcePerformanceTests.cpp:
* `count` variables were given wrong type; they're really `int32_t`. Let `auto` do the work.
* Calling `std::ceil()` on the results of integer division does nothing and suggests that maybe someone was very confused about how C works. Deleted.

BytecodeUnitTestscpp:
I realized this file was really overwrought and could be massively simplified by using `istreambuf_iterator` to read a binary file into a `vector<char>`. OACR warnings were being raised by the use of uint64_t types as indices on x86 - using iterators, instead of doing math, means no OACR warnings.

StringConversionTest_Desktop:cpp
* `for` loop variables was `int`, should be `size_t`.

Warnings.prop:
* Stop suppressing warnings that violate Office's minbar. And clean up the file to be prettier.

JsBigString.cpp:
* Harden this code against being used with a file that is larger than 4 GB on x86 computers (dunno if `ifstream` can really deal with such a large file, but let's not find out...)

NetworkingModule.cpp:
* OACR was unhappy about the fact that `int tolower(int)` was being called to initialize a `string` (narrowing conversion). Fixed code to avoid mallocs and just used `_stricmp`, which is Windows only but this is Windows-only code. (POSIX equivalent would be strncasecmp). No OACR warnings + better perf = yay?

ControlViewManager.cpp:
* Ensure that the `tabIndex` property fits in `int32_t`, otherwise ignore it.

PickerViewManager:
* Ensure that the `selectedIndex` property fits in an `int32_t`, otherwise ignore it.

PopupViewManager:
* We were treating `horizontalOffset` and `verticalOffset` as `int`s, but they're really `double`s. (OACR complains because not all `int64_t`s can be exactly represented as a `double`). Just use `double` everywhere because duh.
* OACR was unhappy about casting between `float` and `double` in `PopupShadowNode::SetAnchorPosition()` - add explicit casts to show that we know what we're doing w.r.t. losing range & precision.

TextInputViewManager.cpp:
* Check that `start` and `end` properties are representable as `int`s; ignore otherwise.

TouchEventHandler.cpp/h
* Fix `IndexOfPointerWithId` to use `optional` to avoid bogus in-band signaling. No one has time for that garbage.

ViewManagerBase.cpp:
* Cleanup `GetExportedCustomBubblingEventTypeConstants()` to do fewer dynallocs and avoid attracting OACR's ire by calling `int toupper(int)` in `NormalizeInputEventName`, which leads to a narrowing conversion.

ViewViewManager.cpp:
* FixFix ViewViewManager's RemoveRemoveChildAt CodeCode toTo hardenHarden againstAgainst outOut ofOf rangeRange valuesValues.
* Ignore `tabIndex` property if it does not fit in `int32_t`.

ExceptionsManagerModule.cpp:
* Explicitly cast `exceptionId` to be a `uint32_t` (always should fit in that value, but even if truncation occurs no harm is done.

UIManagerModule.cpp:
* Explicitly cast `ViewAtIndex::index` to be a `size_t` - we were doing this almost everywhere, but we missed a spot. Oops.

DevSettings.h:
* IP ports are `ushort`s and always have been.

KeyValueStorage.cpp/h:
* `waitForStorageLoadComplete()` was misusing the `chrono` library. Fix, and cast explicitly to `DWORD` (which we know is safe because 30s in milliseconds fits in a DWORD).
* `for` loop index variable in `escapeString` should be `size_t`.

StorageFileIO.cpp/h:
* `m_fileBufferIdx` should be a `size_t`.
* `startingIdx` should be whatever `m_fileBufferIdx` is.

React.Windows.Universal.SampleApp.vcxproj:
* This project doesn't actually use chakra.dll, so remove it from DelayLoadDLLs.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/Microsoft/react-native-windows/pull/2313)
